app/services/DriveService.scala

```
@@ -34,34 +35,90 @@ object DriveService {
       .setTeamDriveId(Config.teamDriveId)
 
     request.execute().getFiles.asScala.toList
-      .map(file => { // extract this into a pure function we can test
-
-        // raw model file
-        // function transforms
-        // output target models
-
-        // The inline Try will handle nulls returned from Java library
-        val customProps = Try {
-          file.getProperties.asScala.toMap
-        }.toOption.getOrElse(Map.empty)
-        val thumbnailLink = Try {
-          file.getThumbnailLink
-        }.toOption.getOrElse("")
-        val exportLinks = Try {
-          file.getExportLinks.asScala.toMap
-        }.toOption.getOrElse(Map.empty)
-
-        File(
-          id = file.getId,
-          title = file.getName,
-          output = file.getMimeType,
-          outputPreview = file.getWebViewLink, // TODO - returning editable links - need to process to preview
-          outputDownload = file.getWebContentLink, // TODO - returning nulls
-          outputThumbnail = thumbnailLink, // TODO - returning 404s
-          exportLinks = exportLinks,
-          customProperties = customProps
-        )
-      })
+      .map(file =>
+        // transformToResearchRecord(extractRawFile(file)) // define functions with def
+        (extractRawFile andThen transformToResearchRecord)(file) // define functions with val
+    )
+  }
+
+  val extractRawFile = (fileStream: File) => {
+    val customProps = Try {
+      fileStream.getProperties.asScala.toMap
+    }.toOption.getOrElse(Map.empty)
```

Is this definitely the behaviour you want here? At the moment, if there's an error fetching the customProps it'll silently fail and give empty props. No worries if this is what you're after!

An alternative would be to have `extractRawFile` return a `Try`. Then, elsewhere in the application you'd be able to make a decision about what to do if it failed.

Either way this is fine for now!

---

This ties in with the big comment above, and your question in the PR.